contributingGuides/philosophies/PAGINATION.md

```diff
@@ -0,0 +1,113 @@
+        3. While they're looking at the middle of the list, more than 1 page (50 items) of data is added to the front of the list. Let's say these items are items 50-150.
+        4. They jump back to the front of the page, and fetch the page at the front, 100-150.
+            - _Note:_ Fetching "all the data they missed" generally isn't a scalable solution, because it's unbounded, and the unbounded loading of data is the problem pagination seeks to solve.
+        5. Now there is a gap between items 51-100 :boom:
+    - "Over-eager eviction"
+        1. A user scrolls far back in the list. So far that we decide there's too much data for us to handle all at once. Performance degrades, and we decide to start evicting the data at the front that's less-recently viewed. Let's say that the user now has items 50-100 loaded, and we've evicted items 101-200.
+        2. New data appears at the front of the list, say items 201-250. We add these items to the list as we normally would if they're looking at the front of the list.
+        3. Now there is a gap between items 101-200 (the data we evicted) :boom:
+
+    If it turns out you _do_ need a strategy for gap detection, here's a high-level summary of how you'd handle it:
+
+    1. Start keeping track of the pages you've loaded in a dedicated Onyx key for pagination (e.g., `report_123_actions_pages`). This is a _sorted_ list of ranges.
+    2. Post-process network responses using the [Pagination middleware](https://github.com/Expensify/App/blob/1a06fa4add10b53a1a9266927d3b08a4ca35d3c4/src/libs/Middleware/Pagination.ts) to keep track of the start and end point of the page you loaded in a request, and [merge it with existing pages if it overlaps](https://github.com/Expensify/App/blob/1a06fa4add10b53a1a9266927d3b08a4ca35d3c4/src/libs/PaginationUtils.ts#L104).
+    3. When rendering your list, use that pages key for your sorting order, and insert "gap markers" between the edges of the pages you've loaded.
+    4. When rendering your list, [only render a single continuous chunk](https://github.com/Expensify/App/blob/1a06fa4add10b53a1a9266927d3b08a4ca35d3c4/src/libs/PaginationUtils.ts#L166) containing your current "anchor point" (the reportAction you linked to, for example), up until you reach the end of the list in either direction or a gap marker.
+    5. Then when you scroll, you'll hit your gap marker and can make network requests to fill in the gap.
+
+    More details can be found in [the Pagination middleware](https://github.com/Expensify/App/blob/1a06fa4add10b53a1a9266927d3b08a4ca35d3c4/src/libs/Middleware/Pagination.ts). Efforts were made to generalize this code, but so far it has only been used for reportActions.
+
+## Future improvements
+The following ideals have not (yet) been implemented in Expensify, but we'll hold them out here as aspirational goals for future pagination systems, inspired by WhatsApp.
```

Comment:
I like the insights here, but a part of me is 👎 on adding aspirational things to a doc meant to explain the current best practices. Feels like something that should go in a P/S or a design doc and then eventually we can update this. For that reason, I'd suggest deleting everything below this comment.

Reply:
I don't feel too strongly about it. I see your point but I also think this gives _some_ small guidance into the future and that we don't necessarily want to do anything that would really hinder us from going in this direction.

A lot of our original philosophies were very aspirational at the time, like the server pushing data to a client's disk. In the early stages, we were only doing that for reportActions and all other data was loaded with GET requests. Now, we let the server push any and all data to clients via `Onyx.update()`.

Reply:
We could address that concern by using some strong language and mentioning specifically that we should never move forward with any solution that involves throwing away data on disk since a future solution might rely on it and point to the slack discussion.

My main concern is that we are including specifics about what we might want to build without a clearly defined problem. Putting solutions before problems and getting too attached to specific solutions are typically discouraged at the org level and I think we should model those values here.